### ISSUE TITLE
[FG:InPlacePodVerticalScaling] Emit a events when resize status changes

### ISSUE THREAD

**Issue by tallclair** [kubernetes/kubernetes#127172]
/kind feature

To help with debugging in place pod resize, the Kubelet should emit events along with various resize status changes:

1. Resize accepted: report the resource deltas
2. Resize infeasible: which resources were over capacity
3. Resize deferred: which resources were over available capacity
5. Resize completed
6. Resize error

/sig node
/priority important-longterm
/milestone v1.32
/triage accepted

**Comment by googs1025** [comment]
/assign

I'm interested in this topic, can I take it on?

**Comment by dshebib** [comment]
Events on resizing errors is waiting to be merged #125654

**Comment by drewhagen** [comment]
Hello @dshebib @googs1025 @tallclair 

This issue has not been updated for a long time, so I'd like to check what's the status. If there's anything we can do, please let us know.

Friendly reminder that code freeze is starting `02:00 UTC Friday November 8th 2024` (**about 3 weeks from now**), and while there is still time, we want to ensure that each PR has a chance to be merged on time.

Is this issue still planned for release `1.32`? Thanks!

**Comment by drewhagen** [comment]
👋 @dshebib @googs1025 @tallclair 
Following up: this issue still hasn't been updated in a while, though I see this is tagged for `v1.32` release and there's some active worked in linked PRs. Is the plan still to resolve this issue for `v1.32`?

Just a reminder that the code freeze is starting `02:00 UTC Friday November 8th 2024` (a little less than **_1 week from now_**). Please make sure any PRs have both `lgtm` and `approved` labels before the code freeze. Thanks!

**Comment by drewhagen** [comment]
👋 Hello @dshebib @googs1025 @tallclair,
Is the plan still to resolve this issue for v1.32?

If so, a gentle reminder that the code freeze has started `02:00 UTC Friday November 8th 2024`. Please make sure any PRs have both `lgtm` and `approved` labels ASAP, and file an [Exception](https://github.com/kubernetes/sig-release/blob/master/releases/EXCEPTIONS.md).
Thanks!

**Comment by pacoxu** [comment]
/milestone v1.32
Add back the milestone:
- sorry, I miss the exception mail https://groups.google.com/g/kubernetes-sig-node/c/OJc4jxLbSaY

**Comment by esotsal** [comment]
> @tallclair I disagree that events help to debug kubelet issues, and in the other side, create bugs or cause a lot of load on the control plan. Do we really need this? Is people that debug issues in prod really using events for debugging and troubleshooting kubelet issues? Events have ttls so they will not used for RCA for sure

I agree, even worse if used in tests will end up on Flaky tests which will be difficult to debug as discussed also [here](https://github.com/kubernetes/kubernetes/issues/114971#issuecomment-1377958562)

**Comment by drewhagen** [comment]
👋 Hello folks @dims @esotsal @pacoxu @aojea  @dshebib @tallclair @googs1025 

Should this issue's priority be changed to `priority/important-soon` as opposed to `priority/important-longterm`?
If not, I propose that we remove the 1.32 milestone so that Release Signal team can focus on higher priority and release blocker issues.

Thanks!

**Comment by sftim** [comment]
I suggest adding the resize completed event; it's the one that other parties are most likely to care about - and it's handy if `kubectl describe` can handle it specially.

**Comment by sftim** [comment]
Having Events available will help us make the tutorial docs that I think we'd prefer to have before https://github.com/kubernetes/enhancements/issues/1287 graduates to GA.

**Comment by tallclair** [comment]
I was thinking that we didn't need events any more with the switch to using conditions to represent the resize status, but having an event for resize completed makes sense (the point at which the in-progress condition is removed?)

**Comment by googs1025** [comment]
> I was thinking that we didn't need events any more with the switch to using conditions to represent the resize status, but having an event for resize completed makes sense (the point at which the in-progress condition is removed?)

cc @shiya0705

**Comment by shiya0705** [comment]
> > I was thinking that we didn't need events any more with the switch to using conditions to represent the resize status, but having an event for resize completed makes sense (the point at which the in-progress condition is removed?)
> 
> cc [@shiya0705](https://github.com/shiya0705)

@googs1025 
Yes, I think so, there are two reasons:
- If Resize status is set as ResizeDeferred (because there is no available resource),  ResizeDeferred event is print, but after node has available resource, Resize will be completed, but no message to indicate it.
- In pod description, limit and request cpu/memory are updated according to the patch command，this does not mean that resize has been successfully applied. Complete event will be helpful for user to make next its next operation.

**Comment by shiya0705** [comment]
Add pod resize complete event: https://github.com/kubernetes/kubernetes/pull/130387

**Comment by wendy-ha18** [comment]
Hello @shiya0705 @tallclair @sftim
Thanks for actively working on this PR.

Friendly reminder that code freeze is starting at 02:00 UTC Friday 21st March 2025 (about **_4 weeks from now_**), and while there is still time, we want to ensure that each PR has a chance to be merged on time.

Please make sure the PR has both `lgtm` and `approved` labels before the code freeze. Thanks!

**Comment by tallclair** [comment]
@sftim Some questions raised on https://github.com/kubernetes/kubernetes/pull/130387 about the resize completed event that I'd be interested in your perspective on:
1. What should the contents of the event message be? The PR is proposing dumping the allocated CPU & memory resources for all containers, but I'm worried that message can get quite large. Maybe with better formatting it wouldn't be so bad though.
2. For a resize that doesn't change the actual resources at all, should the resize completed event be emitted? I think this will only happen if a Deferred or Infeasible resize is reverted

**Comment by sftim** [comment]
If it's feasible to include the extra information (eg current Pod limits and requests) in an annotation, maybe do that. Events can have annotations IIRC.
But I wouldn't block a merge on covering that part. I don't recommend putting that level of detail into `note`.

> For a resize that doesn't change the actual resources at all, should the resize completed event be emitted?

I don't have an opinion, but it would be nice if the event `reason` can make it clear whether a resize actually happened.

**Comment by wendy-ha18** [comment]
Hello @shiya0705 @tallclair @sftim

Appreciate all of your efforts with this PR!

A gentle reminder that the code freeze has started [02:00 UTC Friday 21st March 2025](https://everytimezone.com/s/2c5e9275) . Please make sure this PR has both `lgtm` and `approved` labels ASAP, and file an [Exception](https://github.com/kubernetes/sig-release/blob/master/releases/EXCEPTIONS.md) if you haven't done it yet.
Thanks!

**Comment by Rajalakshmi-Girish** [comment]
Hello @shiya0705 @tallclair @sftim, 

This issue has not been updated for long time, so I'd like to check what's the status. If there's anything we can do, please let us know.

The code freeze is starting 02:00 UTC Friday 25th July 2025 (**about 4 weeks from now**). Please make sure the PR has both lgtm and approved labels before the code freeze. Thanks!

**Comment by tallclair** [comment]
I'd like to get this in to v1.34. I'll prioritize PR reviews for it.

**Comment by shiya0705** [comment]
I'll update the code as soon as the refactoring is complete.
